File: app/lib/llm.ts
import type { SqlDatabase } from "../../db/sql";

/**
 * OpenAI 호환 Chat Completions 어댑터.
 * base URL과 모델명만 바꾸면 OpenAI, Gemini(호환 엔드포인트), DeepSeek, Qwen,
 * Groq, OpenRouter를 같은 코드로 호출합니다.
 */
export type LlmConfig = { baseUrl: string; apiKey: string; model: string; timeoutMs: number };

const DEFAULT_TIMEOUT_MS = 25_000;

function timeoutMs() {
  return Math.max(5_000, Math.min(60_000, Number(process.env.SIGNALIST_LLM_TIMEOUT_MS) || DEFAULT_TIMEOUT_MS));
}

/** DB에 공급자가 하나도 없을 때만 쓰는 환경변수 폴백. */
export function llmConfig(): LlmConfig | null {
  const apiKey = process.env.SIGNALIST_LLM_API_KEY ?? process.env.SIGNALIST_OPENAI_API_KEY;
  if (!apiKey) return null;
  const baseUrl = (process.env.SIGNALIST_LLM_BASE_URL ?? "https://api.openai.com/v1").replace(/\/+$/, "");
  const model = process.env.SIGNALIST_LLM_MODEL ?? process.env.SIGNALIST_OPENAI_NEWS_MODEL ?? "gpt-5.6-terra";
  return { baseUrl, apiKey, model, timeoutMs: timeoutMs() };
}

function extractJson(text: string) {
  const fenced = text.match(/```(?:json)?\s*([\s\S]*?)```/);
  const body = (fenced?.[1] ?? text).trim();
  const start = body.indexOf("{");
  const end = body.lastIndexOf("}");
  if (start < 0 || end <= start) throw new Error("모델이 JSON을 반환하지 않았습니다.");
  return JSON.parse(body.slice(start, end + 1)) as unknown;
}

export class LlmError extends Error {
  constructor(message: string, readonly kind: FailureKind, readonly status = 0) {
    super(message);
    this.name = "LlmError";
  }
}

export type FailureKind = "한도 초과" | "인증 오류" | "일시 오류" | "요청 오류";

/**
 * 실패를 종류별로 나눈다. 한도 초과와 인증 오류는 같은 키로 다시 시도해도 소용이
 * 없으므로 길게 쉬게 하고, 일시 오류만 짧게 쉬었다가 다시 쓴다.
 */
export function classifyFailure(status: number, detail: string): FailureKind {
  if (status === 429) return "한도 초과";
  if (status === 401 || status === 403) return "인증 오류";
  if (status === 402) return "한도 초과";
  if (status >= 500 || status === 408 || status === 0) return "일시 오류";
  if (status === 400 && /quota|exhaust|limit|billing|credit/i.test(detail)) return "한도 초과";
  return "요청 오류";
}

export async function llmJson(config: LlmConfig, system: string, user: string): Promise<unknown> {
  let response: Response;
  try {
    response = await fetch(`${config.baseUrl}/chat/completions`, {
      method: "POST",
      headers: { Authorization: `Bearer ${config.apiKey}`, "Content-Type": "application/json" },
      body: JSON.stringify({
        model: config.model,
        messages: [{ role: "system", content: system }, { role: "user", content: user }],
        response_format: { type: "json_object" },
        temperature: 0.2,
      }),
      signal: AbortSignal.timeout(config.timeoutMs),
    });
  } catch (reason) {
    const message = reason instanceof Error && reason.name === "TimeoutError"
      ? `응답이 ${Math.round(config.timeoutMs / 1000)}초를 넘겨 끊었습니다.`
      : reason instanceof Error ? reason.message : "네트워크 오류";
    throw new LlmError(message, "일시 오류");
  }
  if (!response.ok) {
    const detail = (await response.text()).slice(0, 300);
    throw new LlmError(`${response.status} ${detail}`, classifyFailure(response.status, detail), response.status);
  }
  const payload = await response.json() as { choices?: Array<{ message?: { content?: string } }> };
  const content = payload.choices?.[0]?.message?.content;
  if (!content) throw new LlmError("응답에 본문이 없습니다.", "요청 오류", response.status);
  try {
    return extractJson(content);
  } catch (reason) {
    throw new LlmError(reason instanceof Error ? reason.message : "JSON 해석 실패", "요청 오류", response.status);
  }
}

export type ProviderRow = {
  id: number; name: string; base_url: string; model: string; api_key: string;
  priority: number; is_active: boolean; daily_limit: number; used_today: number;
  usage_date_kst: string | null; cooldown_until: number | null;
  last_status: string; last_error: string | null; last_used_at: number | null;
  success_count: number; failure_count: number;
};

export type ProviderAttempt = { name: string; kind: FailureKind | "성공"; detail?: string };

export function dateKst(time = Date.now()) {
  return new Date(time + 9 * 60 * 60 * 1000).toISOString().slice(0, 10);
}

/** 실패 종류별로 얼마나 쉬게 할지. */
export function cooldownUntil(kind: FailureKind, now: number) {
  if (kind === "한도 초과") {
    // 분당 한도인지 일일 한도인지 응답만 보고는 알 수 없다. 한 시간 쉬되 한국시간
    // 자정을 넘기지 않는다. 자정이면 어차피 무료 티어 사용량이 초기화된다.
    const kst = new Date(now + 9 * 60 * 60 * 1000);
    const midnight = Date.UTC(kst.getUTCFullYear(), kst.getUTCMonth(), kst.getUTCDate() + 1) - 9 * 60 * 60 * 1000;
    return Math.min(midnight, now + 60 * 60 * 1000);
  }
  if (kind === "인증 오류") return now + 6 * 60 * 60 * 1000;
  if (kind === "요청 오류") return now + 30 * 60 * 1000;
  return now + 5 * 60 * 1000;
}

/** 지금 쓸 수 있는 공급자만 우선순위 순으로 가져온다. */
export async function availableProviders(db: SqlDatabase, now = Date.now()): Promise<ProviderRow[]> {
  const rows = await db.prepare(`SELECT id, name, base_url, model, api_key, priority, is_active, daily_limit,
      used_today, usage_date_kst, cooldown_until, last_status, last_error, last_used_at, success_count, failure_count
      FROM llm_providers WHERE is_active = true AND api_key <> ''
      ORDER BY priority, id`).all<ProviderRow>();
  const today = dateKst(now);
  // DB에 공급자가 하나도 없을 때만 환경변수 설정을 마지막 수단으로 끼워 넣는다.
  if (!rows.results.length) {
    const fallback = llmConfig();
    if (!fallback) return [];
    return [{
      id: 0, name: "환경변수", base_url: fallback.baseUrl, model: fallback.model, api_key: fallback.apiKey,
      priority: 999, is_active: true, daily_limit: 0, used_today: 0, usage_date_kst: null, cooldown_until: null,
      last_status: "대기", last_error: null, last_used_at: null, success_count: 0, failure_count: 0,
    }];
  }
  return rows.results.filter((row) => {
    if (row.cooldown_until && row.cooldown_until > now) return false;
    if (!row.daily_limit) return true;
    // 날짜가 바뀌었으면 사용량은 0부터 다시 센다.
    if (row.usage_date_kst !== today) return true;
    return row.used_today < row.daily_limit;
  });
}

async function recordSuccess(db: SqlDatabase, provider: ProviderRow, now: number) {
  if (!provider.id) return;
  const today = dateKst(now);
  await db.prepare(`UPDATE llm_providers SET
      used_today = CASE WHEN usage_date_kst = ? THEN used_today + 1 ELSE 1 END,
      usage_date_kst = ?, cooldown_until = NULL, last_status = '정상', last_error = NULL,
      last_used_at = ?, success_count = success_count + 1, updated_at = ?
    WHERE id = ?`).bind(today, today, new Date(now), new Date(now), provider.id).run();
}

async function recordFailure(db: SqlDatabase, provider: ProviderRow, kind: FailureKind, message: string, now: number) {
  if (!provider.id) return;
  const today = dateKst(now);
  const until = cooldownUntil(kind, now);
  await db.prepare(`UPDATE llm_providers SET
      used_today = CASE WHEN usage_date_kst = ? THEN used_today + 1 ELSE 1 END,
      usage_date_kst = ?, cooldown_until = ?, last_status = ?, last_error = ?,
      last_used_at = ?, failure_count = failure_count + 1, updated_at = ?
    WHERE id = ?`)
    .bind(today, today, new Date(until), kind, message.slice(0, 400), new Date(now), new Date(now), provider.id).run();
}

export type ChainResult = { parsed: unknown | null; provider: ProviderRow | null; attempts: ProviderAttempt[] };

/**
 * 우선순위가 높은 공급자부터 호출하고, 한도 초과·인증 오류·일시 오류가 나면
 * 해당 공급자를 쉬게 한 뒤 다음 공급자로 넘어간다. 전부 실패하면 null이다.
 */
export async function llmJsonWithFailover(db: SqlDatabase, system: string, user: string, now = Date.now()): Promise<ChainResult> {
  const providers = await availableProviders(db, now);
  const attempts: ProviderAttempt[] = [];
  for (const provider of providers) {
    const config: LlmConfig = { baseUrl: provider.base_url.replace(/\/+$/, ""), apiKey: provider.api_key, model: provider.model, timeoutMs: timeoutMs() };
    try {
      const parsed = await llmJson(config, system, user);
      await recordSuccess(db, provider, now);
      attempts.push({ name: provider.name, kind: "성공" });
      return { parsed, provider, attempts };
    } catch (reason) {
      const kind = reason instanceof LlmError ? reason.kind : "일시 오류";
      const message = reason instanceof Error ? reason.message : "알 수 없는 오류";
      await recordFailure(db, provider, kind, message, now);
      attempts.push({ name: provider.name, kind, detail: message.slice(0, 160) });
    }
  }
  return { parsed: null, provider: null, attempts };
}

File: backend/news/crawler/llm.py
"""
OpenAI 호환 Chat Completions 어댑터. `app/lib/llm.ts` 의 순수 부분 이식.

base URL 과 모델명만 바꾸면 OpenAI · Gemini(호환 엔드포인트) · DeepSeek · Qwen · Groq ·
OpenRouter 를 같은 코드로 부른다. DB 를 만지는 페일오버는 `news/services/llm.py` 에 있다.
"""

from __future__ import annotations

import json
import os
import re
import time
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta

import httpx

DEFAULT_TIMEOUT_MS = 25_000
MIN_TIMEOUT_MS = 5_000
MAX_TIMEOUT_MS = 60_000
KST_OFFSET = timedelta(hours=9)
# 503·네트워크 오류는 Gemini 같은 외부 공급자의 순간 혼잡일 수 있다. 한 번의 실패로
# 공급자를 쿨다운시키기 전에 같은 요청을 짧게 다시 보낸다.
TRANSIENT_RETRY_DELAYS_SECONDS = (1.0, 2.0)

_FENCED = re.compile(r"```(?:json)?\s*(.*?)```", re.DOTALL)
_QUOTA_HINT = re.compile(r"quota|exhaust|limit|billing|credit", re.IGNORECASE)


class FailureKind:
    """문자열 상수로 둔다 — DB 의 `last_status` 에 그대로 저장되고 화면에 노출된다."""

    QUOTA = "한도 초과"
    AUTH = "인증 오류"
    TRANSIENT = "일시 오류"
    REQUEST = "요청 오류"
    SUCCESS = "성공"


@dataclass(frozen=True, slots=True)
class LlmConfig:
    base_url: str
    api_key: str
    model: str
    timeout_ms: int


class LlmError(RuntimeError):
    def __init__(self, message: str, kind: str, status: int = 0) -> None:
        super().__init__(message)
        self.kind = kind
        self.status = status


def timeout_ms() -> int:
    raw = os.environ.get("SIGNALIST_LLM_TIMEOUT_MS")
    try:
        value = int(raw) if raw else 0
    except ValueError:
        value = 0
    return max(MIN_TIMEOUT_MS, min(MAX_TIMEOUT_MS, value or DEFAULT_TIMEOUT_MS))


def env_config() -> LlmConfig | None:
    """DB 에 공급자가 하나도 없을 때만 쓰는 환경변수 폴백."""
    api_key = os.environ.get("SIGNALIST_LLM_API_KEY") or os.environ.get(
        "SIGNALIST_OPENAI_API_KEY"
    )
    if not api_key:
        return None
    base_url = (
        os.environ.get("SIGNALIST_LLM_BASE_URL") or "https://api.openai.com/v1"
    ).rstrip("/")
    model = (
        os.environ.get("SIGNALIST_LLM_MODEL")
        or os.environ.get("SIGNALIST_OPENAI_NEWS_MODEL")
        or "gpt-5.6-terra"
    )
    return LlmConfig(base_url, api_key, model, timeout_ms())


def extract_json(text: str) -> object:
    """
    코드펜스로 감싸 오거나 설명을 덧붙이는 모델이 많다. 펜스를 먼저 벗기고, 그 안에서
    첫 `{` 와 마지막 `}` 사이를 잘라 파싱한다.
    """
    fenced = _FENCED.search(text)
    body = (fenced.group(1) if fenced else text).strip()
    start = body.find("{")
    end = body.rfind("}")
    if start < 0 or end <= start:
        raise LlmError("모델이 JSON을 반환하지 않았습니다.", FailureKind.REQUEST)
    return json.loads(body[start : end + 1])


def classify_failure(status: int, detail: str) -> str:
    """
    한도 초과와 인증 오류는 같은 키로 다시 시도해도 소용이 없어 길게 쉬게 하고, 일시
    오류만 짧게 쉬었다가 다시 쓴다.
    """
    if status == 429:
        return FailureKind.QUOTA
    if status in (401, 403):
        return FailureKind.AUTH
    if status == 402:
        return FailureKind.QUOTA
    if status >= 500 or status in (408, 0):
        return FailureKind.TRANSIENT
    if status == 400 and _QUOTA_HINT.search(detail):
        return FailureKind.QUOTA
    return FailureKind.REQUEST


def cooldown_until(kind: str, now: datetime) -> datetime:
    """
    실패 종류별로 얼마나 쉬게 할지.

    한도 초과가 분당 제한인지 일일 제한인지는 응답만 보고 알 수 없다. 한 시간 쉬되
    한국시간 자정을 넘기지 않는다 — 자정이면 어차피 무료 티어 사용량이 초기화된다.
    """
    if kind == FailureKind.QUOTA:
        kst = now + KST_OFFSET
        next_kst_midnight = datetime(
            kst.year, kst.month, kst.day, tzinfo=UTC
        ) + timedelta(days=1)
        midnight = next_kst_midnight - KST_OFFSET
        return min(midnight, now + timedelta(hours=1))
    if kind == FailureKind.AUTH:
        return now + timedelta(hours=6)
    if kind == FailureKind.REQUEST:
        return now + timedelta(minutes=30)
    return now + timedelta(minutes=5)


def llm_json(client: httpx.Client, config: LlmConfig, system: str, user: str) -> object:
    """한 공급자에게 JSON 응답을 요구한다.

    408·5xx·네트워크 오류는 같은 공급자에 짧은 지수 백오프 재시도를 한 뒤에만 호출자에게
    전달한다. 인증·형식 오류는 재시도해도 해결되지 않으므로 즉시 전달한다.
    """
    for delay in (*TRANSIENT_RETRY_DELAYS_SECONDS, None):
        try:
            response = client.post(
                f"{config.base_url}/chat/completions",
                headers={
                    "Authorization": f"Bearer {config.api_key}",
                    "Content-Type": "application/json",
                },
                json={
                    "model": config.model,
                    "messages": [
                        {"role": "system", "content": system},
                        {"role": "user", "content": user},
                    ],
                    "response_format": {"type": "json_object"},
                    "temperature": 0.2,
                },
                timeout=config.timeout_ms / 1000,
            )
            if response.status_code >= 400:
                detail = response.text[:300]
                raise LlmError(
                    f"{response.status_code} {detail}",
                    classify_failure(response.status_code, detail),
                    response.status_code,
                )

            try:
                payload = response.json()
            except ValueError as reason:
                raise LlmError(
                    "응답이 JSON 이 아닙니다.", FailureKind.REQUEST, response.status_code
                ) from reason

            choices = payload.get("choices") or []
            content = (choices[0].get("message") or {}).get("content") if choices else None
            if not content:
                raise LlmError(
                    "응답에 본문이 없습니다.", FailureKind.REQUEST, response.status_code
                )

            try:
                return extract_json(content)
            except LlmError:
                raise
            except ValueError as reason:
                raise LlmError(
                    f"JSON 해석 실패: {reason}", FailureKind.REQUEST, response.status_code
                ) from reason
        except httpx.TimeoutException as reason:
            seconds = round(config.timeout_ms / 1000)
            failure = LlmError(
                f"응답이 {seconds}초를 넘겨 끊었습니다.", FailureKind.TRANSIENT
            )
            failure.__cause__ = reason
        except httpx.HTTPError as reason:
            failure = LlmError(str(reason) or "네트워크 오류", FailureKind.TRANSIENT)
            failure.__cause__ = reason
        except LlmError as reason:
            failure = reason

        if failure.kind != FailureKind.TRANSIENT or delay is None:
            raise failure
        time.sleep(delay)

    raise AssertionError("재시도 루프를 빠져나왔습니다.")
